text/0000-mergeable-rustdoc-cross-crate-info.md

```diff
@@ -0,0 +1,468 @@
+
+* Partial source file index for generating `doc/src-files.js`.
+* Partial search index for generating `doc/search-index.js`.
+* Crate name for generating `doc/crates.js`.
+* Crate name and information for generating `doc/index.html`.
+* Trait implementation list for generating `doc/trait.impl/**/*.js`.
+* Type implementation list for generating `doc/type.impl/**/*.js`.
+* The file may include versioning information intended to assist in generating error messages if an incompatible `doc.parts` is provided through `--include-parts-dir`.
+* The file may contain other information related to cross-crate information that is added in the future.
+
+## New flag: `--parts-out-dir=<path/to>/doc.parts/<crate-name>`
+
+When this flag is provided, the unmerged parts for the current crate will be written to `path/to/doc.parts/<crate name>`. A typical argument is `./target/doc.parts/rand`.
+
+Crates `--include-parts-dir`ed will not appear in `doc.parts`, as `doc.parts` only includes the CCI parts for the current crate.
+
+If this flag is not provided, no `doc.parts` will be written.
+
+The output generated by this flag may be consumed by a future invocation to rustdoc that provides `--include-parts-dir=<path/to>/doc.parts/<crate-name>`.
+
+## New flag: `--include-parts-dir=<path/to/doc.parts/crate-name>`
+
+If this flag is provided, rustdoc will expect that a previous invocation of rustdoc was made with `--parts-out-dir=<path/to/doc.parts/crate-name>`. It will append the parts from the previous invocation to the ones it will render in the doc root (`--out-dir`). The info that's included is not written to its own `doc.parts`, as `doc.parts` only holds the CCI parts for the current crate.
+
+This flag is similar to `--extern-html-root-url` in that it only needs to be provided for externally documented crates. The flag `--extern-html-root-url` controls hyperlink generation. The hyperlink provided in `--extern-html-root-url` never accessed by rustdoc, and represents the final destination of the documentation. The new flag `--include-parts-dir` tells rustdoc where to search for the `doc.parts` directory at documentation-time. It must not be a URL.
+
+In the Guide-level explanation, for example, crate `i` needs to identify the location of `s`'s parts. Since they could be located in an arbitrary directory, `i` must be instructed on where to fetch them. In this example, `s`'s parts happen to be in `./s/target/doc.parts/s`, so rustdoc is called with `--include-parts-dir=s/target/doc.parts/s`.
+
+## New flag: `--include-rendered-docs=<path/to/target/doc/extern-crate-name>`
+
+Rustdoc will assume that `<path/to/target/doc>` was used as the `--out-dir` for `<extern-crate-name>`. This documentation will be copied into the directory specified by `--out-dir`. Rustdoc will effectively run `cp -r <path/to/target/doc/extern-crate-name> <current --out-dir>`.
```

Comment:
I'd really prefer not to add it. As you yourself say, users will often have more complex needs for copying the docs, and there's no advantage other a tiny convenience for rustdoc to include this functionality. The only people using these flags will be implementers of build systems or users who have very complex build requirements. Also, they'll be manually controlling the location and structure of the build directories. So I don't see the downside in requiring them to use a simple `cp` here. If we include the feature, there's also the risk for feature creep where users will want us to add more complex functionality over time.

Note that unlike `--include-parts-dir`, there's no rustdoc-specific behavior. That flag needs to exist because it affects rustdoc's behavior and knowledge of the workspace.

Reply:
I think these are good arguments! @GuillaumeGomez had a valid concern about relying on external tools. However, even without `--include-rendered-docs`, you would only need to rely on external tools for merging in the case of multiple `--out-dir`s, which is an extreme scenario. Having multiple `--out-dir`s is not necessary to avoid having multiple build actions target the same files, since rustdoc writes item docs under a separate directory anyway. And this RFC is mostly about providing options to avoid appending to files within a single `--out-dir`. Even though the multiple `--out-dir` case is anticipated by this RFC, if someone is deciding to split their rustdoc builds in arbitrary ways they should expect to do some manual work.